keyleak/archive_scanner.py

```
@@ -0,0 +1,110 @@
+def _safe_extract_tar(tf: tarfile.TarFile, dest: Path) -> None:
+    dest = dest.resolve()
+    for member in tf.getmembers():
+        target = (dest / member.name).resolve()
+        if not str(target).startswith(str(dest)):
+            raise ArchiveScanError(f"Path traversal in tar entry: {member.name}")
+    # Python 3.12+: pass filter='data' for additional safety. We resolve paths
+    # ourselves above; the filter argument is a defense-in-depth.
+    try:
+        tf.extractall(dest, filter="data")
+    except TypeError:
+        tf.extractall(dest)
```

Ruff (0.15.14)

[error] Uses of `tarfile.extractall()`

(S202)